nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,387 @@
+    # MIT License
+    # https://github.com/SNAS-Series/SNAS-Series
+
+    def __init__(self, total_epochs, curr_temp, base_temp, temp_min=0.33, last_epoch=-1):
+        self.curr_temp = curr_temp
+        self.base_temp = base_temp
+        self.temp_min = temp_min
+        self.last_epoch = last_epoch
+        self.total_epochs = total_epochs
+        self.step(last_epoch + 1)
+
+    def step(self, epoch=None):
+        return self.decay_whole_process()
+
+    def decay_whole_process(self, epoch=None):
+        if epoch is None:
+            epoch = self.last_epoch + 1
+        self.last_epoch = epoch
+        self.total_epochs = 150
+        self.curr_temp = (1 - self.last_epoch / self.total_epochs) * (self.base_temp - self.temp_min) + self.temp_min
+        if self.curr_temp < self.temp_min:
+            self.curr_temp = self.temp_min
+        return self.curr_temp
+
+
+class SNASModule(DartsModule):
+    """
+    The SNAS Module. This is a darts-based method. It uses gumble-softmax to simulate a one-hot distribution to
+    select only one path a time. The SNAS Module should be trained with ParallelTrainValDataLoader in
+    nn.retiarii.oneshot.utils.
+    See base class for more attributes.
+
+    Parameters
+    ----------
+    base_model : pl.LightningModule
+        The module in evaluators in nni.retiarii.evaluator.lightning. User defined model
+        is wrapped by base_model, and base_model will be wrapped by this model.
+    gumble_temperature : float
+        The initial temperature used in gumble-softmax. See [snas] for details.
+    use_temp_anneal : bool
+        True if use a scheduler to anneal gumble temperature.
+    temp_min : float
+        The minimal temperature for scheduler.
+    custom_replace_dict : Dict[Type[nn.Module], Callable[[nn.Module], nn.Module]], default = None
+        The custom xxxChoice replace method. Keys should be xxxChoice type and values should
+        return an nn.module. This custom replace dict will override the default replace
+        dict of each NAS method.
+
+    Reference
+    ----------
+    ..[snas] S. Xie, H. Zheng, C. Liu, and L. Lin, “SNAS: stochastic neural architecture search,” presented at the
+        International Conference on Learning Representations, Sep. 2018. Available: https://openreview.net/forum?id=rylqooRqK7
+    """
+    def __init__(self, base_model, max_epochs, gumble_temperature = 1., use_temp_anneal = False, temp_min = .33, custom_replace_dict=None):
+        # SNAS replace layers require gumble_temperature to work, so self.temp is set before initialize base class
+        self.temp = gumble_temperature
+        super().__init__(base_model, custom_replace_dict)
+        self.use_temp_anneal = use_temp_anneal
+        if use_temp_anneal:
+            self.temp_lr_scheduler = Temp_Scheduler(max_epochs, self.temp, self.temp, temp_min)
+
+    def on_epoch_start(self):
+        if self.use_temp_anneal:
+            self.temp = self.temp_lr_scheduler.step()
```

Comment:
I think we can simply calculate the temperature here. We don't need an extra `TempScheduler`, the calculation is very simple (two lines of code).